documentation/ingestion_server/guides/migrate.md

```diff
@@ -0,0 +1,245 @@
+
+Our goal is to break down an API-involved change into multiple small, atomic
+changes with each step affecting at most one of the ingestion server or the API
+and ensuring that the API and ES remain compatible throughout the process.
+
+## Pull request guidelines
+
+A change that involves modification to the ES index as well as its usage in the
+API requires at least three steps, each associated with exactly one PR that
+modifies exactly one of the ingestion server or the API to allow them to be
+deployed independently.
+
+1. Change the ES index mapping in the ingestion server. Ensure that the change
+   is purely additive, keeping the old fields unchanged and creating new fields
+   that contain the data the API will need.
+
+   This PR should only make changes within the `ingestion_server/` directory.
+
+2. Change the ES fields referenced by the API to use the new fields added in the
+   previous step. Ensure that the old fields become unreferenced.
+
+   The PR should only make changes within the `api/` directory.
+
+3. Change the ES index mapping in the ingestion server to remove the old,
+   now-unreferenced fields.
+
+   This PR should only make changes within the `ingestion_server/` directory.
+
+```{tip}
+Get the PRs reviewed in advance so that the entire process has been vetted by
+the team and there are no surprises or delays when the plans have been set into
+motion.
+```
+
+```{caution}
+Each PR in the chain should branch from, and point to, its predecessor in the
+chain so that CI continues to pass for each PR.
+```
+
+### Example
+
+Assume we have a field `foo` with type `text` in the index. It has a subfield
+`keyword` with type `keyword`. The API uses `foo.keyword` for all purposes.
+
+One PR that does these 3 things would be an API-involved change. So we split
+them into 3 PRs.
+
+1. Changing `foo` to type `keyword` would be an API-free change because it is a
+   type change between two compatible types and does not affect the nested field
+   `foo.keyword` that is in use by the API. Technically the outer field can be
+   assumed to be "new" because it was not being used at all.
+
+2. Then we make an API change to use `foo` directly instead of `foo.keyword`.
+   Any other accommodations to make use of `foo` can be made in this step. In
+   this case `foo` will be the same as `foo.keyword` so no other changes will be
+   needed.
+
+3. Removal of the `foo.keyword` field would now also be an API-free change
+   because the field would no longer be in use.
+
+## Migration process
+
+The entire migration process can be classified into 3 phases.
+
+```{mermaid}
+flowchart TD
+  subgraph api[API]
+    API
+  end
+
+  subgraph elasticsearch[Elasticsearch]
+    image --> image-old
+    image-filtered --> image-old-filtered
+    audio --> audio-old
+    audio-filtered --> audio-old-filtered
+  end
+
+  API --> image
+  API --> image-filtered
+  API --> audio
+  API --> audio-filtered
+```
+
+### Create the new fields
```

Comment:
I think we also need the rollback steps here.

Reply:
Step 2 inside this section points to the `upgrade.md` file which contains the forward steps as well as the rollback steps. Can you elaborate more to help me understand what change you are recommending?

Reply:
Sorry I didn't realize that. It was much easier to understand this in the rendered docs :)